base/range.jl

```diff
@@ -1480,3 +1480,111 @@ julia> mod(3, 0:2)  # mod(3, 3)
 """
 mod(i::Integer, r::OneTo) = mod1(i, last(r))
 mod(i::Integer, r::AbstractUnitRange{<:Integer}) = mod(i-first(r), length(r)) + first(r)
+
+
+"""
+    LogRange(start, stop, length)
+
+Construct a specialized array whose elements are spaced logarithmically
+between the given endpoints. That is, the ratio of successive elements is
+a constant, calculated from the length.
+
+Like [`LinRange`](@ref), floating-point error may cause it to miss intermediate
+rational values, but the first and last elements should be exactly those provided.
+
+This is similar to `geomspace` in Python. Unlike `PowerRange` in Mathematica,
+you specify the number of elements not the ratio.
+Unlike `logspace` in Python and Matlab, the `start` and `stop` arguments are
+always the first and last elements of the result, not powers applied to some base.
+
+!!! compat "Julia 1.9"
+    This function requires at least Julia 1.9.
+
+# Examples
+```jldoctest
+julia> LogRange(10, 4000, 3)
+3-element LogRange{Float64}:
+ 10.0, 200.0, 4000.0
+
+julia> ans[2] ≈ sqrt(10 * 4000)  # middle element is the geometric mean
+true
+
+julia> LogRange(-62_500, -0.01, 5) |> collect  # endpoints are always exact
+5-element Vector{Float64}:
+ -62500.0
+  -1250.0
+    -25.0
+     -0.49999999999999994
+     -0.01
+
+julia> LogRange(√2, 32, 10)
+10-element LogRange{Float64}:
+ 1.41421, 2.0, 2.82843, 4.0, 5.65685, 8.0, 11.3137, 16.0, 22.6274, 32.0
+
+julia> LogRange(1, 1000, 4) ≈ 10 .^ (0:3)  # was logspace(0,3,4) in Julia < 1.0
+true
+
+julia> LogRange(1, -1+0im, 5) ≈ cis.(LinRange(0, pi, 5))  # complex numbers
+true
+```
+"""
+struct LogRange{T} <: AbstractArray{T,1}
+    start::T
+    stop::T
+    len::Int
```

Comment:
LinRange has a length of parametric type, unless there is a reason LogRange should differ we should do the same here.

Reply:
That was changed in https://github.com/JuliaLang/julia/pull/41619 , after this PR.

The main goal was StepRangeLen, because other changes caused these to be produced in cases where StepRange was previously used, and some people wanted a StepRange of BigInt:
```julia
julia> StepRange(1,2,3) * 4 |> typeof
StepRange{Int64,Int64}                    # julia 1.0
StepRangeLen{Int64, Int64, Int64, Int64}  # now, to allow for zero step
```
I'm not entirely sure why LinRange was changed there, since floating point ranges longer than `Int` seem hard to imagine. But perhaps someone may convert a StepRange to a LinRange and want this to always work -- every AbstractRange can be converted to LinRange, perhaps you want to standardise there for some purpose.

Is there any conceivable use for a LogRange longer than Int? There is no path to inadvertantly convert a StepRange etc.
```julia
julia> StepRange(big(1),2,big(3)) |> length |> typeof
BigInt

julia> exp.(StepRange(big(1),2,big(3))) |> length |> typeof
Int64
```
If not, I'm inclined to say we should prefer the simplicity of `Int` over the symmetry to AbstractRange types. (And probably standardise indices on `Int` too, as suggested below.)

One concern besides simplicity is whether the guarantee that the endpoints will always be exact still holds. I don't have a failure case, but BigInt length will produce things like this as intermediate steps:
```julia
julia> 0.01 ^ (big(1) / 1)
0.01000000000000000020816681711721685132943093776702880859375
```

Reply:
I can't think of a reasonable example of where a type larger other than Int would be appropriate right now.